moshi/src/main/java/com/squareup/moshi/JsonUtf8Reader.java

```diff
@@ -329,6 +325,13 @@ private int doPeek() throws IOException {
       } else {
         checkLenient();
       }
+    } else if (peekStack == JsonScope.STREAMING_VALUE) {
+      valueSource.discard();
+      valueSource = null;
+      stackSize--;
+      pathIndices[stackSize - 1]++;
+      pathNames[stackSize - 1] = "null";
+      return doPeek();
```

Comment:
... now it should happen less frequently. I accomplish this by using the stack to hint that a value source is non-null.